icepyx/core/visualization.py

```diff
@@ -0,0 +1,503 @@
+                        bbox_ij[2] = lonmax
+
+                    if bbox_ij[3] > latmax:
+                        bbox_ij[3] = latmax
+
+                    bbox_list.append(bbox_ij)
+
+            return bbox_list
+
+        else:
+            return [self.bbox]
+
+    def query_icesat2_filelist(self) -> tuple:
+        """
+        Query list of ICESat-2 files for each bounding box
+
+        Returns
+        -------
+        filelist_tuple : tuple
+            A tuple of non-empty lists of bounding boxes and corresponding ICESat-2 file lists
+        """
+        # a list of 5*5 bounding boxes
+        bbox_list = self.grid_bbox()
+
+        is2_bbox_list = []
+        is2_file_list = []
+
+        for bbox_i in bbox_list:
+
+            region = ipx.Query(
+                self.product,
+                bbox_i,
+                self.date_range,
+                cycles=self.cycles,
+                tracks=self.tracks,
+            )
+            icesat2_files = region.avail_granules(ids=True)[0]
+
+            if not icesat2_files:
+                continue
+            else:
+                all_cycles = list(set(region.avail_granules(cycles=True)[0]))
+                icesat2_files_latest_cycle = files_in_latest_n_cycles(
+                    icesat2_files, [int(c) for c in all_cycles]
+                )
+                is2_bbox_list.append(bbox_i)
+                is2_file_list.append(icesat2_files_latest_cycle)
+
+        is2_file_tuple = zip(is2_bbox_list, is2_file_list)
+
+        return is2_file_tuple
+
+    def generate_OA_parameters(self) -> list:
+        """
+        Get metadata from file lists in each 5*5 bounding box.
+
+        Returns
+        -------
+        paras_list : list
+            A list of parameters for OpenAltimetry API query, including the
+            reference ground track (RGT), cycle number, datetime, bounding box,
+            and product name.
+        """
+        # list of parameters for API query
+        paras_list = []
+
+        # tuple of non-empty bbox and ICESat-2 datafiles
+        filelist_tuple = self.query_icesat2_filelist()
+
+        for bbox_i, filelist in filelist_tuple:
+
+            for fname in filelist:
+
+                if self.product in ["ATL07", "ATL10"]:
+                    format_string = (
+                        "{product:5d}-{HH:02d}_{datetime:%Y%m%d%H%M%S}_{rgt:04d}{cycle:02d}{"
+                        "region:02d}_{release:03d}_{version:02d}.h5"
+                    )
+                else:
+                    format_string = (
+                        "{product:5d}_{datetime:%Y%m%d%H%M%S}_{rgt:04d}{cycle:02d}{"
+                        "region:02d}_{release:03d}_{version:02d}.h5"
+                    )
+                temp = intake.source.utils.reverse_format(
+                    format_string=format_string, resolved_string=fname
+                )
+
+                rgt = temp["rgt"]  # Reference Ground Track
+                cycle = temp["cycle"]  # Cycle number
```

Comment:
As I was looking through `visualization.py` to try and add some more tests, it occurred to me that this is similar to some code Tyler added for querying granules by cycle/track. I'll dig into it more tomorrow, but on a quick look it seems like we could use (`granules.gran_IDs()`)[https://github.com/icesat2py/icepyx/blob/4d6e64b95e61609364385e080195c39c4486b182/icepyx/core/granules.py#L36] to get this info here.

Reply:
Yes looks like so, will take a closer look and test the function!

Reply:
OK, added a new function `gran_paras` based on `granules.gran_IDs()` to get these information: https://github.com/icesat2py/icepyx/pull/144/commits/60c6d8a6157feb1db1a3d057d1341562bb652650

Reply:
Awesome! I took it a step further and wrapped your function into `granules.gran_IDs()` directly so that we should only have to maintain the string parsing in one location. I think that long-term we could probably streamline this type of string parsing/handling across all modules, but I don't think we should hold this PR up any more for that sort of thing.

Reply:
Yes, put the date string in one place looks so much better!